### ISSUE TITLE
Errorneous matching of the immediately following block command after @ref

### ISSUE THREAD

**Issue by mosra** [doxygen/doxygen#6715]
Similarly to #6689, this is a regression caused by @albert-github's #6592 (commit https://github.com/doxygen/doxygen/commit/07ae32a61e779647bace1efdbf8d15b35871fcca). Processing the following snippet:

```cpp
/** This is a @ref Foo @htmlonly yay! @endhtmlonly */
struct Foo {};
```

will produce this warning:

```
a.cpp:1: warning: unable to resolve reference to `Foo@htmlonly' for \ref command
a.cpp:1: warning: htmlonly section ended without end marker
a.cpp:1: warning: unexpected command endhtmlonly
```

This happens with other block-like commands such as `@xmlonly`, `@verbatim` or `@code` (and probably others) placed immediately after the `@ref` argument. I'm hitting especially the `@xmlonly` / `@htmlonly` case, for example in the docs for [Magnum::Animation::Easing](https://doc.magnum.graphics/magnum/namespaceMagnum_1_1Animation_1_1Easing.html), where I'm using it to add custom CSS-styled elements and inline SVG to the output. Another usual case is `@code` that's hit very often by users of the [m.css Doxygen theme](https://mcss.mosra.cz/doxygen/) for inline highlighted code snippets.

Possible workarounds:

- reverting 07ae32a61e779647bace1efdbf8d15b35871fcca, which makes this go away completely
- adding an explicit newline after each affected `@ref` (might not be wanted in some contexts)
- using `@link name @endlink` instead of `@ref name` (unnecessarily verbose)

**Comment by albert-github** [comment]
In the documentation it is stated:

> 24.106 \ref <name> ["(text)"]
> 
> Creates a reference to a named section, subsection, page or anchor. For HTML documentation the reference command will generate a link to the section. For a section or subsection the title of the section will be used as the text of the link. For an anchor the optional text between quotes will be used or <name> if no text is specified. For LaTeX documentation the reference command will generate a section number for sections or the text followed by a page number if <name> refers to an anchor.

I didn't look into the code yet, but probably in case no double quote is found the first word is taken as the third (optional) argument. Probably a fix can be to exclude commands from the start of the start of the third argument of `\ref`, paying special attention to escaped commands i.e `@@`, `\\` etc.

**Comment by mosra** [comment]
Just to clarify: this worked very well and without problems with 1.8.13, 1.8.14 all the way until https://github.com/doxygen/doxygen/commit/07ae32a61e779647bace1efdbf8d15b35871fcca, which adds `{FILEMASK}` to `REFWORD` (and reverting this particular bit makes it working again).

`{FILEMASK}` includes `@` and `\`, while neither `@` nor `\` was matched before.

A completely different idea: disallowing `@ref` to take complex filenames as arguments and allowing these only in `@link`? That's already the case for links to `operator""` (see my PR #6713), where `@ref` would get confused by the quotes, so one has to use `@link` instead.

**Comment by albert-github** [comment]
I think the `operator""` case is a quite different case and could probably work when using using single quotes around the `<text>` argument.
Seen the `FILEMASK` remark my remark about excluding commands will not work as the characters 
`@` and especially the `\` are valid `FILEMASK` characters. A `\` is on Windows a possible start character of a directory path and (Windows users) use all types of characters at the beginning of a file name. The syntax states '\ref ["(text)"]' so this looks like mandatory double quotes to me and in case the are not present there will be no (automatic) third argument. This has to be checked and maybe enforced (looking into the fix won't break other working things). Another solution might be that the `FILEMASK` in this case is restricted to filenames not staring with a `@`, the `\` will be hard to exclude as it is a valid path separator.

**Comment by mosra** [comment]
> could probably work when using single quotes around the <text> argument

I even have a half-written test case and a bug report for this from 2014, but halfway in I realized the added complexity (and potential bug surface) would not be worth it, so I didn't submit anything. For the record:

```cpp
/**
 * @page operator Operators
 *
 * See also @ref operator "" _page_.
 */

/**
 * @brief Page literal
 *
 * See also the @ref operator "Operators" _page_.
 */
Page operator "" _page_(unsigned long long int);
```

> Another solution might be that the FILEMASK in this case is restricted to filenames not staring with a `@`, the `\` will be hard to exclude as it is a valid path separator.

I would dare to say:

- A slash at the start is mainly used for absolute paths on Unix platforms, in which case it's a forward slash (`/`), not `\`, absolute paths on Windows (where backslash matters) are usually starting with drive letter (except for e.g. `\\Server\Volume\File`).
- I can't think of *any* use case where `@ref` to an *absolute* path (or network location, even) would make sense -- and saying e.g. `@ref /home/mosra/Code/foo/bar.h` is a bad project smell anyway, no value in supporting that.

**Comment by mosra** [comment]
@doxygen @albert-github Any chance this could get **finally** fixed? You have a repro case and everything here. I have to stay at a forked version from December last year because with this issue (and a bunch of other exciting new bugs that I have to yet report, sigh), 1.8.16 is completely unusable for me.

**Comment by doxygen** [comment]
The commit referenced is a way of fix the problem (by not removing the space between `Foo` and `@htmlonly` during internal processing, so that at the moment `@ref is evaluated` the argument is `Foo` and not `Foo@htmlonly`.

But there is (of course) a drawback I just realized:
`1 @htmlonly 2@endhtmlonly 3` will then produce
`1<space><space>2<space>3` instead of `1<space>2<space>3` as one might expect and this is no way to get only one space. 

There will probably be someone in the world unhappy with this change ☹️

**Comment by albert-github** [comment]
Well there is probably a mean trick:
```
1&zwnj;@htmlonly 2@endhtmlonly 3
```
or
```
1&zwj;@htmlonly 2@endhtmlonly 3
```

**Comment by doxygen** [comment]
@albert-github yes this is an option, but not something a user would (or should) have to think about. 

All things considered, I still prefer to limit the file mask by excluding the `@` character as an allowed character in file names. Note that the `\` does not have to be excluded as a `\htmlonly` command is already converted to `@htmlonly` internally.

**Comment by albert-github** [comment]
I agree the `&zwnj;` is something the user should  not have to think about.

Regarding the disabling the `@` is in my personal opinion a bad sign to have in a filename (together with some other stuff like spaces Windows users tend to put into filenames).
- would it be sufficient to restrict the `FILEMASK`  in respect to the `@` on the first position only?

(There will probably be someone in the, Windows, world unhappy with this change :frowning_face:)

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.8.17.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).
